llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -2904,6 +2924,41 @@ static bool isKnownNonZeroFromOperator(const Operator *I,
       case Intrinsic::vector_reduce_smax:
       case Intrinsic::vector_reduce_smin:
         return isKnownNonZero(II->getArgOperand(0), Depth, Q);
+        // If we know the reduction doesn't overflow and all elements are
+        // non-zero, the reduction is non-zero.
+      case Intrinsic::vector_reduce_mul:
+      case Intrinsic::vector_reduce_add:
+        if (computeKnownBits(I, Depth + 1, Q).isNonZero())
+          return true;
```

Comment:
It is incorrect. `llvm.vector.reduce.mul.i8(<4 x i8> <i8 4, i8 4, i8 4, i8 4>)` evaluates to zero.

Reply:
This is `I`, the result, not the input vector.

Reply:
Oh, sorry for my mistake.

Reply:
BTW, I wouldn't spend your time reviewing this unless we decide to actually try to get it in.